### PR TITLE
Implement trueloc/falseloc

### LINKED ISSUE #568: true_pos
In a Fortran Discourse thread [Indexing arrays by an array of logicals](https://fortran-lang.discourse.group/t/indexing-arrays-by-an-array-of-logicals/2264) I mentioned a `true_pos` function I had written to simulate logical indexing, and @ivan-pi suggested that it be considered for stdlib.  To deal with the case of a `lbound /= 1` mentioned by mecej4, I have added an optional `lower_bound` argument. If `tf(:)` has the same size as `x(:)`, one can write `x(true_pos(tf))` to refer to elements of x where `tf` is `.true.` One can also write `pack(x,tf)`, but that cannot appear on the LHS of an assignment. I have used `true_pos` hundreds of times in my codes -- maybe it would be generally useful. Since Fortran has `minloc`, `maxloc`, and `findloc`, maybe `true_pos` should be renamed `trueloc`.

```
pure function true_pos(tf, lower_bound) result(ipos)
! return in ipos the positions of the true elements in tf(:)
logical, intent(in)           :: tf(:)
integer, intent(in), optional :: lower_bound 
integer                       :: ipos(count(tf))
integer                       :: i,j
j = 0
do i=1,size(tf)
   if (tf(i)) then
      j = j + 1
      ipos(j) = i
   end if
end do
if (present(lower_bound)) ipos = ipos + lower_bound - 1
end function true_pos
```

### PR DESCRIPTION
**Pull request by awvwgk**
Closes #568

### CODE REVIEW

File: src/stdlib_array.f90 [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+contains
+
+  !> Return the positions of the true elements in array
+  pure function trueloc(array, lbound) result(loc)
+    !> Mask of logicals
+    logical, intent(in) :: array(:)
+    !> Lower bound of array to index
+    integer, intent(in), optional :: lbound
+    !> Locations of true elements
+    integer :: loc(count(array))
```
Comment (gareth-nx) [inline src/stdlib_array.f90:19]: I wonder how it will perform if `size(array) == 0`? I suspect it will still work (and return a logical array of length 0), but it would be good to check this and either include such a test in the test-suite, or else document that it won't work.

Comment (gareth-nx) [reply src/stdlib_array.f90:19]: I won't repeat this comment below, but obviously it applies to all the routines

File: doc/specs/stdlib_array.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+#### Syntax
+
+`call [[trueloc(function)]] (array[, lbound])`
+
+#### Arguments
+
+`array`: List of default logical arrays. This argument is `intent(in)`.
+
+`lbound`: Lower bound of the array to index. This argument is `optional` and `intent(in)`.
```
Comment (gareth-nx) [inline doc/specs/stdlib_array.md:34]: Somewhere we should document the `return value`.

File: doc/specs/stdlib_array.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+#### Syntax
+
+`call [[falseloc(function)]] (array[, lbound])`
+
+#### Arguments
+
+`array`: List of default logical arrays. This argument is `intent(in)`.
+
+`lbound`: Lower bound of the array to index. This argument is `optional` and `intent(in)`.
```
Comment (gareth-nx) [inline doc/specs/stdlib_array.md:68]: As above, suggest to document the return value

File: src/tests/array/test_logicalloc.f90 [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      ]
+  end subroutine collect_logicalloc
+
+  subroutine test_trueloc_where(error)
+    !> Error handling
+    type(error_type), allocatable, intent(out) :: error
+
+    integer :: ndim
+    real, allocatable :: avec(:), bvec(:), cvec(:)
+
```
Comment (gareth-nx) [inline src/tests/array/test_logicalloc.f90:33]: Good for the tests to include some corner cases, such as `size(array) == 0`, and also `array(:) = .true.` and `array(:) = .false.`, for both `trueloc` and `falseloc` (unless they do not support that, in which case just document it in the specs).

**Review by gareth-nx — COMMENTED** [review]
Thanks -- a few simple suggestions here, which should be easy to address.

**Review by gareth-nx — APPROVED** [review]
Thanks

**Comment by milancurcic** [comment]
I have a concern about the implementation.

Since `trueloc` and `falseloc` call `logicalloc` and they're all functions, there will be a double copy of the result array on each call of `trueloc` and `falseloc`. It's possible that some of the poor performance shown [here](https://github.com/fortran-lang/stdlib/issues/568#issuecomment-997374204) can be explained by the double copy. So maybe there is a performance penalty in doing it this way.

Why not implement the functionality of `logicalloc` in both `trueloc` and `falseloc` (and duplicate a few lines of code) and avoid the extra function call and array copy?

And, is `falseloc` really needed? You could call `trueloc(.not. condition)` to get that. I understand it's a matter of style and has no right or wrong answer.

FWIW, my preference would be, in this order:

1. Have only `trueloc` and avoid calling `logicalloc`
2. Have `trueloc` and `falseloc` and avoid calling `logicalloc`
3. Current implementation.

**Comment by awvwgk** [comment]
Implementation in 7b5ffac 
```
[Timing] trueloc: 0.0218s where: 0.0043s ratio:  5.0
[Timing] trueloc: 0.0218s merge: 0.0044s ratio:  4.9
[Timing] falseloc: 0.0223s where: 0.0051s ratio:  4.4
[Timing] falseloc: 0.0233s merge: 0.0053s ratio:  4.4
```

Using subroutine instead of function
```
[Timing] trueloc: 0.0181s where: 0.0043s ratio:  4.2
[Timing] trueloc: 0.0168s merge: 0.0044s ratio:  3.8
[Timing] falseloc: 0.0174s where: 0.0050s ratio:  3.5
[Timing] falseloc: 0.0178s merge: 0.0052s ratio:  3.4
```

Duplicated inline implementation
```
[Timing] trueloc: 0.0184s where: 0.0042s ratio:  4.4
[Timing] trueloc: 0.0168s merge: 0.0046s ratio:  3.7
[Timing] falseloc: 0.0177s where: 0.0049s ratio:  3.6
[Timing] falseloc: 0.0177s merge: 0.0053s ratio:  3.3
```

**Comment by jvdp1** [comment]
> Implementation in [7b5ffac](https://github.com/fortran-lang/stdlib/commit/7b5fface9f0556d18248fef61d1e307a9247e0f8)
> 
> ```
> [Timing] trueloc: 0.0218s where: 0.0043s ratio:  5.0
> [Timing] trueloc: 0.0218s merge: 0.0044s ratio:  4.9
> [Timing] falseloc: 0.0223s where: 0.0051s ratio:  4.4
> [Timing] falseloc: 0.0233s merge: 0.0053s ratio:  4.4
> ```
> 
> Using subroutine instead of function
> 
> ```
> [Timing] trueloc: 0.0181s where: 0.0043s ratio:  4.2
> [Timing] trueloc: 0.0168s merge: 0.0044s ratio:  3.8
> [Timing] falseloc: 0.0174s where: 0.0050s ratio:  3.5
> [Timing] falseloc: 0.0178s merge: 0.0052s ratio:  3.4
> ```
> 
> Duplicated inline implementation
> 
> ```
> [Timing] trueloc: 0.0184s where: 0.0042s ratio:  4.4
> [Timing] trueloc: 0.0168s merge: 0.0046s ratio:  3.7
> [Timing] falseloc: 0.0177s where: 0.0049s ratio:  3.6
> [Timing] falseloc: 0.0177s merge: 0.0053s ratio:  3.3
> ```

Nice comparison. However, is this comparison fair for `trueloc`, since the aim is a bit different than `where` and `merge`? Furthermore, `trueloc` must initialize an `integer array, which is not the case of where and merge, if I understand it well.

**Comment by awvwgk** [comment]
> Nice comparison. However, is this comparison fair for `trueloc`, since the aim is a bit different than `where` and `merge`? Furthermore, `trueloc` must initialize an `integer array, which is not the case of where and merge, if I understand it well.

Indeed, `merge` and `where` are both language intrinsic constructs and give the compiler many opportunities for optimization, due to the `elemental` character merge is somewhat slower than `where`. For `trueloc` we have to construct an index array (function call, stack allocation, loop over array and increment counter), which can only be slower, with link time optimization the compiler might be able to optimize `trueloc` away again and regain the performance loss.

I would consider `trueloc` and `falseloc` a convenience feature, for actual compute intense task `where` is preferable (also offers worksharing via OMP).

**Comment by awvwgk** [comment]
Hmm, now I want to have timers in test-drive and stdlib.

**Review by milancurcic — APPROVED** [review]

File: doc/specs/stdlib_array.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+#### Return value
+
+Returns an array of default integer size, with a maximum length of `size(array)` elements.
+
+#### Examples
+
+```fortran
+program demo
```
Comment (jvdp1) [inline doc/specs/stdlib_array.md:43]:
```suggestion
program demo_trueloc
```

File: doc/specs/stdlib_array.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  use stdlib_array, only : trueloc
+  implicit none
+  real, allocatable :: array(:)
+  allocate(array(500))
+  call random_number(array)
+  array(trueloc(array > 0.5)) = 0.0
+end program demo
```
Comment (jvdp1) [inline doc/specs/stdlib_array.md:50]:
```suggestion
end program demo_trueloc
```

File: doc/specs/stdlib_array.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+Experimental
+
+#### Description
+
+Turn a logical mask into an index array by selecting all true values.
+
+#### Syntax
+
+`call [[trueloc(function)]] (array[, lbound])`
```
Comment (jvdp1) [inline doc/specs/stdlib_array.md:28]: Is it a function of a subroutine?
In the demo, it is a function.

File: doc/specs/stdlib_array.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+Experimental
+
+#### Description
+
+Turn a logical mask into an index array by selecting all false values.
+
+#### Syntax
+
+`call [[falseloc(function)]] (array[, lbound])`
```
Comment (jvdp1) [inline doc/specs/stdlib_array.md:66]: Function or subroutine?

File: doc/specs/stdlib_array.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+`lbound`: Lower bound of the array to index. This argument is `optional` and `intent(in)`.
+
+#### Return value
+
+Returns an array of default integer size, with a maximum length of `size(array)` elements.
+
+#### Examples
+
+```fortran
+program demo
```
Comment (jvdp1) [inline doc/specs/stdlib_array.md:81]:
```suggestion
program demo_falseloc
```

File: doc/specs/stdlib_array.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  use stdlib_array, only : falseloc
+  implicit none
+  real, allocatable :: array(:)
+  allocate(array(-200:200))
+  call random_number(array)
+  array(falseloc(array < 0.5), lbound(array)) = 0.0
+end program demo
```
Comment (jvdp1) [inline doc/specs/stdlib_array.md:88]:
```suggestion
end program demo_falseloc
```

File: doc/specs/stdlib_array.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+## Procedures and methods provided
+
+
+### `trueloc`
+
+#### Status
+
+Experimental
+
+#### Description
```
Comment (jvdp1) [inline doc/specs/stdlib_array.md:22]: Mention somewhere that it is a pure function.

File: src/stdlib_array.f90 [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  implicit none
+  private
+
+  public :: trueloc, falseloc
+
+contains
+
+  !> Return the positions of the true elements in array
+  pure function trueloc(array, lbound) result(loc)
+    !> Mask of logicals
```
Comment (jvdp1) [inline src/stdlib_array.f90:14]:
```suggestion
    !! Version: experimental
    !!
    !! [Specification](LINK_FOR_FORD)
    !> Mask of logicals
```

**Review by jvdp1 — CHANGES_REQUESTED** [review]
Thahnk you.
Here are some minor comments

File: doc/specs/stdlib_array.md (hunk 8 above)
Comment (awvwgk) [reply doc/specs/stdlib_array.md:66]:
```suggestion
`loc = [[falseloc(function)]] (array[, lbound])`
```

File: doc/specs/stdlib_array.md (hunk 7 above)
Comment (awvwgk) [reply doc/specs/stdlib_array.md:28]:
```suggestion
`loc = [[trueloc(function)]] (array[, lbound])`
```

**Comment by gareth-nx** [comment]
> I would consider `trueloc` and `falseloc` a convenience feature, for actual compute intense task `where` is preferable (also offers worksharing via OMP).

I suggest that something like this is stated in the specs to help guide users.

**Comment by gareth-nx** [comment]
Looks like this is good to be merged? @jvdp1, any further comments?

**Review by jvdp1 — APPROVED** [review]
LGTM. Thank you @awvwgk
